openpgp-dsm/src/lib.rs

```diff
@@ -80,24 +80,98 @@ pub struct DsmAgent {
     role:        Role,
 }
 
-const DSM_LABEL_PGP:       &str = "sq_dsm";
-const ENV_API_KEY:         &str = "FORTANIX_API_KEY";
-const ENV_API_ENDPOINT:    &str = "FORTANIX_API_ENDPOINT";
-const ENV_APP_UUID:        &str = "FORTANIX_APP_UUID";
-const ENV_HTTP_PROXY:      &str = "http_proxy";
-const ENV_NO_PROXY:        &str = "no_proxy";
-const ENV_P12:             &str = "FORTANIX_PKCS12_ID";
-const MIN_DSM_VERSION:     &str = "4.2.0";
+/// The version of this crate.
+pub const SQ_DSM_VERSION: &str = env!("CARGO_PKG_VERSION");
+const DSM_LABEL_PGP:      &str = "sq_dsm";
+const ENV_API_KEY:        &str = "FORTANIX_API_KEY";
+const ENV_API_ENDPOINT:   &str = "FORTANIX_API_ENDPOINT";
+const ENV_APP_UUID:       &str = "FORTANIX_APP_UUID";
+const ENV_HTTP_PROXY:     &str = "http_proxy";
+const ENV_NO_PROXY:       &str = "no_proxy";
+const ENV_P12:            &str = "FORTANIX_PKCS12_ID";
+const MIN_DSM_VERSION:    &str = "4.2.0";
 // As seen on sdkms-client-rust/blob/master/examples/approval_request.rs
-const OP_APPROVAL_MSG:     &str = "This operation requires approval";
+const OP_APPROVAL_MSG:    &str = "This operation requires approval";
 
 #[derive(Clone)]
-enum Auth {
+pub enum Auth {
     ApiKey(String),
     // App UUID and PKCS12 identity
     Cert(Uuid, Identity),
 }
 
+impl Auth {
+    pub fn from_options_or_env(
+        cli_api_key: Option<&str>,
+        cli_client_cert: Option<&str>,
+        cli_app_uuid: Option<&str>,
+    ) -> Result<Self> {
+        // Try API key
+        let api_key = match (cli_api_key, env::var(ENV_API_KEY).ok()) {
```

Comment:
Thanks, I didn't know about that. We have three scenarios that could error, according to the doc:

- The var is not set -> we need to discard the error and look in the CLI params
- The var name has a "=" or NULL -> Will never happen since they are hardcoded
- The var value is not unicode -> We could specifically check if the error is `VarError::NotUnicode` but I don't think it's really needed, the user will get a "no auth credentials found" error and will inspect the variables themselves.

Reply:
My assumption is that the env vars are still set by the user? The current handling should cause issue when env is set and at the same time `--api-key` is used. If the env fails due to error of unicode and `--api-key` is used without informing user.

But I agree this is a rare and made-up scenario. We could choose to handle it but even the current code is fine. I don't have any preference here, okay with whatever you choose.